src/Lightweight/DataMapper/DataMapper.hpp

```diff
@@ -463,6 +469,593 @@ class DataMapper: public std::enable_shared_from_this<DataMapper>
+
+        Reflection::EnumerateMembers(firstRecord, [reader = &reader]<size_t I, typename Field>(Field& field) mutable {
+            if constexpr (IsField<Field>)
+            {
+                if constexpr (Field::IsOptional)
+                    field.MutableValue() = reader->GetNullableColumn<typename Field::ValueType::value_type>(I + 1);
+                else
+                    field.MutableValue() = reader->GetColumn<typename Field::ValueType>(I + 1);
+            }
+            else if constexpr (SqlGetColumnNativeType<Field>)
+            {
+                if constexpr (Field::IsOptional)
+                    field = reader->GetNullableColumn<typename Field::BaseType>(I + 1);
+                else
+                    field = reader->GetColumn<Field>(I + 1);
+            }
+        });
+
+        Reflection::EnumerateMembers(secondRecord, [reader = &reader]<size_t I, typename Field>(Field& field) mutable {
+            if constexpr (IsField<Field>)
+            {
+                if constexpr (Field::IsOptional)
+                    field.MutableValue() = reader->GetNullableColumn<typename Field::ValueType::value_type>(
+                        Reflection::CountMembers<FirstRecord> + I + 1);
+                else
+                    field.MutableValue() =
+                        reader->GetColumn<typename Field::ValueType>(Reflection::CountMembers<FirstRecord> + I + 1);
+            }
+            else if constexpr (SqlGetColumnNativeType<Field>)
+            {
+                if constexpr (Field::IsOptional)
+                    field =
+                        reader->GetNullableColumn<typename Field::BaseType>(Reflection::CountMembers<FirstRecord> + I + 1);
+                else
+                    field = reader->GetColumn<Field>(Reflection::CountMembers<FirstRecord> + I + 1);
+            }
+        });
+    }
+
+    template <typename Record>
+    bool ReadSingleResult(SqlServerType sqlServerType, SqlResultCursor& reader, Record& record)
+    {
+        auto const outputColumnsBound = CanSafelyBindOutputColumns<Record>(sqlServerType);
+
+        if (outputColumnsBound)
+            BindAllOutputColumns(reader, record);
+
+        if (!reader.FetchRow())
+            return false;
+
+        if (!outputColumnsBound)
+            GetAllColumns(reader, record);
+
+        return true;
+    }
+} // namespace detail
+
+template <typename Record, typename Derived>
+
+inline SqlCoreDataMapperQueryBuilder<Record, Derived>::SqlCoreDataMapperQueryBuilder(DataMapper& dm,
+                                                                                     std::string fields) noexcept:
+    _dm { dm },
+    _formatter { dm.Connection().QueryFormatter() },
+    _fields { std::move(fields) }
+{
+}
+
+template <typename Record, typename Derived>
+size_t SqlCoreDataMapperQueryBuilder<Record, Derived>::Count()
+{
+    _dm.Statement().ExecuteDirect(_formatter.SelectCount(this->_query.distinct,
```

Comment:
```suggestion
    auto stmt = SqlStatement { _dm.Connection() };
    stmt.ExecuteDirect(_formatter.SelectCount(this->_query.distinct,
```
do not expose `SqlStatement`'s internal `SqlStatement` instance, but use it like above (also for the other cases)